prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -317,25 +317,68 @@ int64_t rxr_pkt_req_cq_data(struct rxr_pkt_entry *pkt_entry)
 	return cq_data_hdr->cq_data;
 }
 
-size_t rxr_pkt_req_max_header_size(int pkt_type)
+
+size_t rxr_pkt_calc_data_size(struct rxr_ep *ep,
+		              fi_addr_t addr,
+			      int pkt_type,
+			      uint64_t fi_flags,
+			      size_t rma_iov_count)
 {
-	/* max_hdr_size does not include optional connid hdr length because
-	 * it is impossible to have both optional connid hdr and opt_raw_addr_hdr
-	 * in the header, and length of opt raw addr hdr is larger than
-	 * connid hdr (which is confirmed by the following assertion).
-	 */
-	assert(RXR_REQ_OPT_RAW_ADDR_HDR_SIZE >= sizeof(struct rxr_req_opt_connid_hdr));
+       struct rdm_peer *peer;
+       uint16_t header_flags = 0;
 
-	int max_hdr_size = REQ_INF_LIST[pkt_type].base_hdr_size
-		+ RXR_REQ_OPT_RAW_ADDR_HDR_SIZE
-		+ sizeof(struct rxr_req_opt_cq_data_hdr);
+       peer = rxr_ep_get_peer(ep, addr);
+       assert(peer);
 
-	if (pkt_type == RXR_EAGER_RTW_PKT ||
-	    pkt_type == RXR_DC_EAGER_RTW_PKT ||
-	    pkt_type == RXR_LONGCTS_RTW_PKT)
-		max_hdr_size += RXR_IOV_LIMIT * sizeof(struct fi_rma_iov);
+       if (peer->is_local) {
+                assert(ep->use_shm);
+                return rxr_env.shm_max_medium_size;
+       }
 
-	return max_hdr_size;
+       if (rxr_peer_need_raw_addr_hdr(peer))
+            header_flags |= RXR_REQ_OPT_RAW_ADDR_HDR;
+       else if (rxr_peer_need_connid(peer))
+            header_flags |= RXR_PKT_CONNID_HDR;
+
+       if (fi_flags & FI_REMOTE_CQ_DATA)
+            header_flags |= RXR_REQ_OPT_CQ_DATA_HDR;
+
+       return ep->mtu_size - rxr_pkt_req_max_header_size_extended(pkt_type,
+							           header_flags,
+								   rma_iov_count);
+}
+
+
+size_t rxr_pkt_req_max_header_size_extended(int pkt_type, uint16_t flags, size_t rma_iov_count)
+{
+       int max_hdr_size = REQ_INF_LIST[pkt_type].base_hdr_size;
+
+       if(flags & RXR_REQ_OPT_RAW_ADDR_HDR) {
+                /* It is impossible to have both optional connid hdr and opt_raw_addr_hdr
+                 * in the header, and length of opt raw addr hdr is larger than
+                 * connid hdr (which is confirmed by the following assertion).
+                 */
+                assert(RXR_REQ_OPT_RAW_ADDR_HDR_SIZE >= sizeof(struct rxr_req_opt_connid_hdr));
+                max_hdr_size += RXR_REQ_OPT_RAW_ADDR_HDR_SIZE;
+       } else if(flags & RXR_PKT_CONNID_HDR)
+            max_hdr_size += sizeof(struct rxr_req_opt_cq_data_hdr);
```

Comment:
Wrong struct here?

Reply:
Yes -- it should be rxr_req_opt_connid_hdr